stan/math/prim/scal/prob/neg_binomial_2_lpmf.hpp

```diff
@@ -79,33 +84,60 @@ return_type_t<T_location, T_precision> neg_binomial_2_lpmf(
   }
 
   for (size_t i = 0; i < max_size_seq_view; i++) {
-    if (include_summand<propto>::value) {
-      logp -= lgamma(n_vec[i] + 1.0);
-    }
-    if (include_summand<propto, T_precision>::value) {
-      logp += multiply_log(phi__[i], phi__[i]) - lgamma(phi__[i]);
-    }
-    if (include_summand<propto, T_location>::value) {
-      logp += multiply_log(n_vec[i], mu__[i]);
-    }
-    if (include_summand<propto, T_precision>::value) {
-      logp += lgamma(n_plus_phi[i]);
-    }
-    logp -= (n_plus_phi[i]) * log_mu_plus_phi[i];
-
-    // if phi is large we probably overflow, defer to Poisson:
-    if (phi__[i] > 1e5) {
-      logp = poisson_lpmf(n_vec[i], mu__[i]);
-    }
-
-    if (!is_constant_all<T_location>::value) {
-      ops_partials.edge1_.partials_[i]
-          += n_vec[i] / mu__[i] - (n_vec[i] + phi__[i]) / (mu__[i] + phi__[i]);
-    }
-    if (!is_constant_all<T_precision>::value) {
-      ops_partials.edge2_.partials_[i]
-          += 1.0 - n_plus_phi[i] / (mu__[i] + phi__[i]) + log_phi[i]
-             - log_mu_plus_phi[i] - digamma(phi__[i]) + digamma(n_plus_phi[i]);
+    if (phi__[i] > internal::neg_binomial_2_phi_cutoff) {
+      // Phi is large, deferring to Poisson.
+      // Copying the code here as just calling
+      // poisson_lpmf does not preserve propto logic correctly
+      if (include_summand<propto>::value) {
+        logp -= lgamma(n_vec[i] + 1.0);
+      }
+      if (include_summand<propto, T_location>::value) {
+        logp += multiply_log(n_vec[i], mu__[i]) - mu__[i];
+      }
+
+      if (!is_constant_all<T_location>::value) {
+        // This is the Taylor series of the full derivative for phi -> Inf
+        // Obtained in Mathematica via
+        // Series[n/mu - (n + phi)/(mu+phi),{phi,Infinity, 1}]
+        // Currently ignoring the term (mu__[i] - n_vec[i]) / phi__[i]
+        ops_partials.edge1_.partials_[i] += n_vec[i] / mu__[i] - 1;
+      }
+
+      // The derivative wrt. phi = 0 + O(1/neg_binomial_2_phi_cutoff^2),
+      // But the quadratic term is big enough to warrant inclusion here
+      // (can be around 1e-6 at cutoff).
+      // Expansion obtained in Mathematica via
+      // Series[1 - (n + phi) / (mu + phi) + Log[phi] - Log[mu + phi] -
+      //  PolyGamma[phi] + PolyGamma[n + phi],{phi,Infinity, 2}]
+      if (!is_constant_all<T_precision>::value) {
+        ops_partials.edge2_.partials_[i]
+            += (mu__[i] * (-mu__[i] + 2 * n_vec[i]) + n_vec[i] * (1 - n_vec[i]))
+               / (2 * square(phi__[i]));
+      }
+
+    } else {
+      if (include_summand<propto, T_precision>::value) {
+        logp += binomial_coefficient_log(n_plus_phi[i] - 1, n_vec[i]);
+      }
+
+      logp += phi__[i] * (log_phi[i] - log_mu_plus_phi[i])
+              - (n_vec[i]) * log_mu_plus_phi[i];
```

Comment:
This doesn't need parens around `n_vec[i]`.

Reply:
Thanks for spotting this, removed.